### ISSUE TITLE
Distribute VS Code via a signed apt repository

### ISSUE THREAD

**Issue by Tyriar** [Microsoft/vscode#2973]
This would make auto-updating possible on Debian-based OS's.

**Comment by Tyriar** [comment]
Unfortunately it doesn't look like the signing will be up in time for the release, pushing to April. /cc @joaomoreno

**Comment by joaomoreno** [comment]
Signatures are always a lengthy process...

It's enough if we get the debs and rpms out the door.

**Comment by typekpb** [comment]
does this one cover rpm repository as well, or should I create a new issue?

**Comment by Tyriar** [comment]
Being pushed to May.

**Comment by Tyriar** [comment]
@typekpb https://github.com/Microsoft/vscode/issues/229 is tracking the rpm/yum repo.

**Comment by Tyriar** [comment]
Still waiting on signing.

**Comment by Tyriar** [comment]
Finally progress is being made on this, I'll be going to a meeting towards the end of the month which should give the way forward :+1: 

As such, it will likely come in July as the meeting occurs during the test/end game week.

**Comment by lijunle** [comment]
Will the repo include the insider builds?

**Comment by Tyriar** [comment]
@lijunle the insiders build will be available via repo yes.

**Comment by kaanoo2904** [comment]
So ... Will the repo be available in July ? It's near the end of month ... :)

**Comment by Tyriar** [comment]
Nope but it should be coming soon, need to be patient unfortunately :frowning:

**Comment by edgard** [comment]
Please consider adding a check on whether we want or not for the apt repository source list to be created, otherwise it would be a really messy to use something like Ansible/Puppet in order to automate installations.

Currently it's just erasing and creating the conf again: https://github.com/Microsoft/vscode/commit/d624636afe4676835b2f313efd884f711599f3b9

Something like this (similar to what chrome does):

```
DEFAULTS_FILE="/etc/default/vscode"

if [ -f "${DEFAULTS_FILE}" ]; then
    return
fi
```

**Comment by tagplus5** [comment]
Auto updated PPA for ubuntu based on official vscode deb's
https://github.com/tagplus5/vscode-ppa

**Comment by artem-sidorenko** [comment]
I'm packaging released tgz in my [OBS repo](https://software.opensuse.org/download.html?project=home%3Aartem_sidorenko&package=vscode) for Ubuntu/Debian. Maybe somebody can use that as intermediate solution

**Comment by sudopluto** [comment]
Is there any word on getting the apt repo signed?

**Comment by carl-di-ortus** [comment]
Add to concerns - any chance to make it until [Debian freeze](https://wiki.debian.org/DebianStretch)? Not sure how this happens, but I believe the "no new packages" is the deadline, namely: 2017-01-05.

**Comment by n00dl3** [comment]
@carl-di-ortus that only concerns official debian packages, not unofficial repositories...

**Comment by Tyriar** [comment]
Update: SQL Server for Linux is now being shipped on the new Microsoft Linux repositories that we've been waiting on. I should have access soon to start integrating it into our build process :tada: 

See https://docs.microsoft.com/en-us/sql/linux/sql-server-linux-setup-ubuntu and https://docs.microsoft.com/en-us/sql/linux/sql-server-linux-setup-red-hat for an idea of what's going to be available.

**Comment by fsateler** [comment]
@Tyriar great news! But please do not use `apt-key add ` in your documentation as well. The better usage is to add the (binary) gpg keyring to `/etc/apt/trusted.gpg.d`:

```sh
curl https://packages.microsoft.com/keys/microsoft.asc | gpg --dearmor > microsoft.gpg
sudo mv microsoft.gpg /etc/apt/trusted.gpg.d/microsoft.gpg
```

**Comment by Tyriar** [comment]
@fsateler any reason that is better? Delegating that to `apt` seems like the best thing to do, what if they change how it's done in the future?

**Comment by fsateler** [comment]
@Tyriar the thing is that `apt-key add` requires habing the full `gnupg` installed. However, if the microsoft keyring is also downloadable as a binary file (ie, dearmored), then only the more lightweight `gpgv` will be used by apt. In fact, newer versions of apt do not depend on `gnupg` but only on `gpgv`. Currently `gnupg` is still Recommended, but that may be demoted further in the future.

**Comment by sudopluto** [comment]
I think for simplicity it is better to let apt handle everything.

**Comment by n00dl3** [comment]
>apt-key add requires having the full gnupg installed (...) In fact, newer versions of apt do not depend on gnupg but only on gpgv

However, there is no point installing a debian(-based) distro without the `apt` package - except if you want to make things harder than they are - and so without a functional `apt-key` command, no matter what dark magic is behind it.

**Comment by fsateler** [comment]
> However, there is no point installing a debian(-based) distro without the apt package

Correct, but that doesn't mean that all commands will work without additional packages. For example, `https` sources will not work without installing `apt-transport-https`. Similarly, `apt-key add` does not work if you do not have `gnupg` installed.

> I think for simplicity it is better to let apt handle everything.

I think dropping a file in a directory (if microsoft can publish the binary gpg keyring) is simpler than running arbitrary `apt` commands.

**Comment by Tyriar** [comment]
> I think dropping a file in a directory (if microsoft can publish the binary gpg keyring) is simpler than running arbitrary apt commands.

While I agree that's simpler, `apt` could be doing or may do additional work in the future. This is one of the main benefits of deferring this to `apt` which almost certainly knows how to install keys in itself than us :smiley:

If you're so inclined to install it manually by downloading the file and moving it manually then you're obviously free to do so.

**Comment by Tyriar** [comment]
I have some more information on the system now, submitting our .deb files to the Microsoft apt repository looks to like it can be completely automated through our build system. Due to the November/December iteration starting to wrap up I'll be too busy wrapping up the planned stuff this iteration so **expect the .deb to start shipping via repo on Insiders sometime in January**. Stable to come soon after if everything goes smoothly.

**Comment by julian-klode** [comment]
We have deprecated adding stuff with apt-key in APT. This might not work, depending on the system (it needs gnupg installed, which we don't want to force on users).

There are two ways to do things:

1. Provide a way to manually add the file to trusted.gpg.d
2. Provide a repository configuration package containing the key (and maybe a sources.list.d file)
3. Do the same thing as in 2, but in each package shipping in the repo (Google style)

If you just have one app Google style is quite nice: You install the deb, and it automatically configures apt to take care of updates.

**Comment by eosrei** [comment]
Other methods are valid, but `apt-key` is the recommended method in Debian/Ubuntu. 

`apt-key` used by:
* Docker - https://docs.docker.com/engine/installation/linux/ubuntulinux/
* Heroku - https://toolbelt.heroku.com/install-ubuntu.sh
* PagerDuty - https://www.pagerduty.com/docs/guides/agent-install-guide/
* Gitlab - https://packages.gitlab.com/gitlab/gitlab-ce/install

**Comment by julian-klode** [comment]
@eosrei  I just told you what the recommended method is. And as the person who does all the apt releases (call me the "apt release manager"), I seem to be more qualified to such a statement than you are.

The other thing is the historical method. We deprecated this a few months ago, so word did not spread very far yet. That said, recent apt-key warns about non-interactive use.

I'm open for another format for repository configuration so we can have some apt add-repository command at some point which downloads some bundle of sources.list entries and a keyring. That needs to be designed, though.

For the immediate future, a repository configuration package is definitely the best approach. Just stuff a <packagename>.gpg file in trusted.gpg.d and a <packagename>.list in sources.list.d if you have per-app repos like google; or for a single repo, provide a package with config for that repo and then tell the user to download that and do:

 apt install repoconfig.deb
 apt update
 apt install package

This really is the most sane approach there is.

**Comment by eosrei** [comment]
Thank you for the details! Your Github profile isn't clear who you are and the "We" is ambiguous. I jumped to a conclusion in defense of an open source maintainer... I will double check in the future.

**Comment by julian-klode** [comment]
@eosrei Right. My profile does not include the bio part. How rude of me. 

And if anyone has proposals for better integration with 3rd party repositories, I'm sure we're all happy to hear them on deity@lists.debian.org. It's clearly suboptimal right now, but we can start improving that in a year or so (freeze hits in two months, no point trying to get something fancy in now).

**Comment by eosrei** [comment]
@julian-klode Haha! Thanks for adding it. :) I wouldn't have questioned you at all.

**Comment by Tyriar** [comment]
@julian-klode thanks for joining the discussion :smile: If installing keys with `apt-key` is deprecated then we will used a different method :+1: 

I was originally considering doing it the "Google style" and installing the key/repo when you first install the package, in fact remnants of what was going to be my impl [still exist](https://github.com/Microsoft/vscode/blob/8ca9455e9ffde36fa4835eda067e3badc0d81c82/resources/linux/debian/postinst.template#L38). I'll probably go in this direction, potentially with a way to disable the key/repo installation via an env var as someone mentioned they didn't like it in the past.

**Comment by twright-msft** [comment]
@julian-klode  - 
I'm on the SQL Server engineering team at Microsoft.

Thanks for the heads up and official clarification on this.  I had heard about this and went to the man pages to check on it.  Starting in 16.10 there is now a warning, but it wasn't as clear to me as saying that it is deprecated and folks should move away from using it.

http://manpages.ubuntu.com/manpages/yakkety/man8/apt-key.8.html

Currently it states:
       Note that if usage of apt-key is desired the additional installation of
       the GNU Privacy Guard suite (packaged in gnupg) is required. For this
       reason alone the programmatic usage (especially in package
       maintainerscripts!) is strongly discouraged. Further more the output
       format of all commands is undefined and can and does change whenever
       the underlying commands change.  apt-key will try to detect such usage
       and generates warnings on stderr in these cases.

From what I can tell, gnupg is installed by default in the versions of Ubuntu that we support (16.04, 16.10), so we have not run into a problem with this yet.

We'll investigate your proposal of using a package to install the key.  

It would be helpful to either update the man pages online or state somewhere more clearly that in fact apt-key will be deprecated.

Again, thanks for the clarification here!

**Comment by julian-klode** [comment]
Yes, it's installed by default for a transitional period (apt still recommends it, and the recommends likely pull it in, or it's explicitly seeded - I don't know about the latter). That's why you don't run into any problems now. Once the Recommends on gnupg turns into a suggest (probably around the 18.04 timeline), you might have an issue.

I think "programmatic usage (especially in package maintainerscripts!) is strongly discouraged" is fairly clear that it's not supposed to be used anymore. Well, not about the "anymore" part, but it should be reasonably sufficient to convey the idea.

**Comment by rickmed** [comment]
hello, any progress? I was excited to update from apt in february but had download again from website 😞

**Comment by Tyriar** [comment]
I'm actively working on this right now, I expect it to be available for insiders within a few days.

**Comment by julian-klode** [comment]
@Tyriar If you want some feedback from the APT perspective feel free to e-mail me.

**Comment by Tyriar** [comment]
Took a little more than a few days...but from today's insiders the new signed apt repository should be installed automatically :tada: The repo is currently 64-bit only and contains both `code` and `code-insiders`, so install one via a `.deb` file and you can install the other through `apt-get`.

You can checkout https://github.com/Microsoft/vscode/blob/master/resources/linux/debian/postinst.template for how the auto install works, basically it installs the key and repo if they don't exist yet. I created this follow up https://github.com/Microsoft/vscode/issues/20785 for re-enabling the repo if it was disabled due to a distro upgrade as the deb package tries to be distro-agnostic.

@julian-klode here's how I'm doing the key install https://github.com/Microsoft/vscode/blob/735a7c0cd72483d1dc2be6f2a227517c39ffdf9d/resources/linux/debian/postinst.template#L41, basically exactly what @fsateler suggested.

If you want to manually install the repo here's how to do it:

```bash
# Install key
curl https://packages.microsoft.com/keys/microsoft.asc | gpg --dearmor > microsoft.gpg
sudo mv microsoft.gpg /etc/apt/trusted.gpg.d/microsoft.gpg
# Install repo
sudo sh -c 'echo "deb [arch=amd64] https://packages.microsoft.com/repos/vscode stable main" > /etc/apt/sources.list.d/vscode.list'
# Update apt-get
sudo apt-get update
# Install
sudo apt-get install code # or code-insiders
```

Please ping this issue if you have any problems, it would be great to work out any kinks that exists particularly in the auto-install before this goes to stable. Enjoy! :beers:

**Comment by ChaseMoskal** [comment]
@Tyriar — vscode is prompting me to update from v0.11.1 to v0.11.2 via [the website](https://code.visualstudio.com/updates/v1_11#_1112-update)

I tried to update with

    sudo apt-get install code

It tells me

    code is already the newest version (1.11.1-1491486998).

I suppose the version that apt-get has access to is out-of-date

**Comment by sudopluto** [comment]
@ChaseMoskal 
to upgrade from apt you have to run
`sudo apt update && sudo apt upgrade`
you are getting that error because you didnt update your sources most likely

**Comment by ghetolay** [comment]
@ChaseMoskal be sure you have a `/etc/apt/sources.list.d/vscode.list` and that the content is not commented (with a `#`).

It happened to me after an Ubuntu upgrade.

**Comment by ChaseMoskal** [comment]
@sudopluto, @ghetolay — it works!

 - I closed everything I had open, and ran `sudo apt update && sudo apt upgrade`

 - I opened `/etc/apt/sources.list.d/vscode.list` and saw that its full contents was:

       deb [arch=amd64] https://packages.microsoft.com/repos/vscode stable main

 - After `sudo apt-get install code` — I have code v**1.11.2**-1492070517

Thanks fellas, I appreciate it

**Comment by Tyriar** [comment]
As @ghetolay mentioned the repo will be disabled after an Ubuntu upgrade, I made a change this morning that works similar to Chrome that will re-enable the repo when the deb package is installed next (coming in tomorrow's Insiders) https://github.com/Microsoft/vscode/issues/24898. This will get auto-updating working again after you install manually after an OS upgrade.

**Comment by ghetolay** [comment]
@Tyriar nice thks!

Rarely seen a PR built out of a comment :+1

**Comment by akashnimare** [comment]
Hey @Tyriar 
I'm working on an electron app named [zulip-electron](github.com/zulip/zulip-electron). I have already build a .deb using [electron-builder](https://github.com/electron-userland/electron-builder). Just wanted to know how to add `.deb` to a PPA so that users can install/update my app via `sudo apt-get myapp`. Any suggestions would be really appreciated.

**Comment by Tyriar** [comment]
@akashnimare we don't use a PPA, we use a regular apt repo and the actual implementation is handled by another team. There should be plenty of documentation on the internet for how to set up a repo.

**Comment by akashnimare** [comment]
So I guess I just need to setup a repo + the above post install script?

**Comment by Tyriar** [comment]
@akashnimare the post install is optional, the minimal thing you need to do is setup a repo and provide users with steps to register it like https://code.visualstudio.com/docs/setup/linux#_debian-and-ubuntu-based-distributions

**Comment by Z-Y00** [comment]
why not provide the "debian/* "  file,  in case someone may wanna pack it into Debian  , this would save much trouble.

**Comment by Tyriar** [comment]
@Z-Y00 I don't understand the question? We provide the deb package via the website homepage and https://code.visualstudio.com/docs/setup/linux#_debian-and-ubuntu-based-distributions ?

**Comment by Z-Y00** [comment]
I mean you can consider geting it into Debian's official repository, since the code is under MIT license.
https://www.debian.org/doc/debian-policy/
